prompts/prompt_template.go

```diff
@@ -0,0 +1,315 @@
+package prompts
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/samber/lo"
+	"github.com/tmc/langchaingo/schema"
+)
+
+var (
+	ErrInputVariableReserved      = errors.New("Cannot have input variable")
+	ErrPartialVariableReserved    = errors.New("Cannot have partial variable")
+	ErrInvalidPartialVariableType = errors.New("Invalid partial variable type")
+)
+
+// PromptTemplateBaseOptions are the options that are common to all prompt templates.
+type PromptTemplateBaseOptions struct {
+	// A list of variable names the prompt template expects
+	InputVariables []string
+	// How to parse the output of calling an LLM on this formatted prompt
+	OutputParser schema.OutputParser[any]
+	// Partial variables
+	PartialVariables map[string]any
+}
+
+// checkInputVariables validates the input variable names do not include restricted names.
+func checkInputVariables(inputVariables []string) error {
+	for _, variable := range inputVariables {
+		if variable == "stop" {
+			return ErrInputVariableReserved
+		}
+	}
+
+	return nil
+}
+
+// checkPartialVariables validates the partial variable names do not include restricted names.
+func checkPartialVariables(partialValues map[string]any) error {
+	for k := range partialValues {
+		if k == "stop" {
+			return ErrPartialVariableReserved
+		}
+	}
+
+	return nil
+}
+
+// applyPromptTemplateBaseOptions applies the input options to the default options.
+func applyPromptTemplateBaseOptions(input PromptTemplateBaseOptions) (PromptTemplateBaseOptions, error) {
+	opts := PromptTemplateBaseOptions{
+		InputVariables:   make([]string, 0),
+		OutputParser:     nil,
+		PartialVariables: make(map[string]any),
+	}
+	if len(input.InputVariables) != 0 {
+		if err := checkInputVariables(input.InputVariables); err != nil {
+			return opts, err
+		}
+		opts.InputVariables = append(opts.InputVariables, input.InputVariables...)
+	}
+	if input.OutputParser != nil {
+		opts.OutputParser = input.OutputParser
+	}
+	if len(input.PartialVariables) != 0 {
+		if err := checkPartialVariables(input.PartialVariables); err != nil {
+			return opts, err
+		}
+
+		for k, v := range input.PartialVariables {
+			opts.PartialVariables[k] = v
+		}
+	}
```

Comment:
Checking against option is less readable for now, do we need to switch to the `WithXXX(...) CallOptions` pattern for better readabilities to override the default options from input options?

I saw your previous comments said we should keep the parameter pattern the same as TypeScript and Python. And I saw some of the PRs.

Reply:
Can you propose an addition of that?